common-pipeline/pipeline-flow/pipeline-flow-v3-schema.json

```diff
@@ -493,6 +517,18 @@
           "description": "optional link name (used in parameter sets when there are multiple input sources)",
           "type": "string"
         },
+        "link_type": {
+          "description": "link type",
+          "enum": [
+            "primary",
+            "secondary",
+            "tertiary"
+          ]
+        },
```

Comment:
Why isn't this stored in the add_data section of the link?

Reply:
From object point of view, link type is a property to link. 
It is optional property. Default is primary.

Reply:
RIght - but why does it have to be in the main part of the schema? We generally put fields in the schema when they are likely to be shared by other applications that use the schema. This seems likely to be a Data Stage specific field especially since you enumerated the values the field can have.

Reply:
It could be shared by other applications. 
With agreement from Curtis / Matt, Primary, Secondary, Tertiary each could represent any type of link as application wants.

Reply:
But if we add application specific fields into the main pipleineFlow schema we would end up with many many fields for each object. The idea of app_data is that it is a holding place for application specific attributes leaving, where possible, the main objects in pipelineFlow to contain the fields that are shared between applications. Even the UI fields have to be stored in `app_data` (and believe me I tried very hard for them to be in the main schema objects). So there needs to be a convincing argument that link_type is required by other applications for it top go in the main link object.

The second point here is the within the current pipelineFlow it is the ports that generally describe the type of data they accept in that they have a label and other attributes that describe that data. It is not clear to me whether you intend to set different `link_type`s on multiple links going into the same port or set different `link_type`s on links going into different ports. What will you do if the user changes this link_type attribute -- wouldn't that mean the label on the port will need to change? You might think the label is unimportant but what if you wanted to surface the port tooltip function in common-canvas which displays the label? You'd then have to manage changing the `link_type` with changing port information and keeping them in sync. And if you're suggesting other apps use link_type they will have to do the same thing.

Reply:
We worked with our architects (Jean-Claude, Kevin Cherkauer, and Ritesh Kumar Gupta) in coming up with a set of properties that we think are useful for other teams. Datastage has many other properties which we are putting them in datastage specific sections. The only properties are proposed here are the ones we think can be common in future. Even though the link type may not be used by other teams now, but we are proposing looking forward how these other models can evolve know the different types of data we need to process day by day.  For datastage these link types are very important because these describe the type of data and it behavior depending on the type of link. For example reject means the data that doesn't conform to the rules of the company policy of application so the properties are more in context of this behavior. When we had discussion with Curtis he suggested that all models may not have the reject but they may have other link types depending on the data they process. So keep it generic he suggested to use primary, secondary and tertiary and each team can interpret the links according to the behavior of the tool. We understand this hides the usability a bit but for the benefit of keeping it generic we decided it's OK to loose that usability.
In Datastage every link will have its input port and output port.  We intend to set different link_types on multiple links going into different ports. For datastage port label has no use. Our operator json describes the cardinality for these different types of links so any stage can accept only certain types of links with min and max. UI can control this when user is adding links. User can also convert one link type to another.   Link type has nothing to do with the port information.  The columns in process through these links won't change unless user chooses to do so.
I hope we answered your questions. If you have further questions I would ike to set up call and discuss.

Reply:
>> Datastage has many other properties which we are putting them in datastage specific sections.

If that's true I don't understand why its so important that we put `link_type` in the link object given that no other dev team has asked for it.

Reply:
Agreed. In general we recommend using app_data unless:
1) The information is needed by backend processors. All front-end information goes into app_data.
2) The information is useful for other consumers of the spec

Reply:
@tomlyn @curtis-browning
The link_type is used by back-end processors -- it's not just UI information. It fundamentally affects how the engine will execute the flow. It also seems like quite a generic concept that other tools might make use of in the future -- it's essentially as basic as an ID, except for the class rather than the instance. Basically we have an object (a link) and an ID for it, but without link_type we don't know what class it is. Seems like very fundamental stuff to me.

I do suggest the existing enumeration is suboptimal -- I would rather see link_type as an integer. Then any tool that uses it can decide whatever values they want and are not going to run out of them as they will with current "primary", "secondary", "tertiary".

Reply:
I would suggest changing `link_type` to a `string`.  This allows for more flexibility in the future if needed and allows teams to use it as they see fit.

Reply:
Agreed. I have changed the link_type to a string.